crates/mev-share-rpc-api/src/auth.rs

```diff
@@ -0,0 +1,148 @@
+//! A layer responsible for implementing flashbots-style authentication
+//! by signing the request body with a private key and adding the signature
+//! to the request headers.
+
+use std::{
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use ethers_core::{types::H256, utils::keccak256};
+use ethers_signers::Signer;
+use futures_util::future::BoxFuture;
+
+use http::{header::HeaderValue, HeaderName, Request};
+use hyper::Body;
+
+use tower::{Layer, Service};
+
+/// Layer that applies [`FlashbotsSigner`] which adds a request header with a signed payload.
+#[derive(Clone)]
+pub struct FlashbotsSignerLayer<S> {
+    signer: Arc<S>,
+}
+
+impl<S> FlashbotsSignerLayer<S> {
+    /// Creates a new [`FlashbotsSignerLayer`] with the given signer.
+    pub fn new(signer: Arc<S>) -> Self {
+        FlashbotsSignerLayer { signer }
+    }
+}
+
+impl<S: Clone, I> Layer<I> for FlashbotsSignerLayer<S> {
+    type Service = FlashbotsSigner<S, I>;
+
+    fn layer(&self, inner: I) -> Self::Service {
+        FlashbotsSigner {
+            signer: self.signer.clone(),
+            inner,
+        }
+    }
+}
+
+/// Middleware that adds a request header with a signed payload.
+/// For more info, see https://docs.flashbots.net/flashbots-auction/searchers/advanced/rpc-endpoint#authentication
+#[derive(Clone)]
+pub struct FlashbotsSigner<S, I> {
+    signer: Arc<S>,
+    inner: I,
+}
+
+impl<S, I> Service<Request<Body>> for FlashbotsSigner<S, I>
+where
+    I: Service<Request<Body>> + Clone + Send + 'static,
+    I::Future: Send,
+    S: Signer + Clone + Send + 'static,
+{
+    type Response = I::Response;
+    type Error = I::Error;
```

Comment:
we need a bit more error handling here to get rid of the unwraps

>  I::Error: Into<Box<dyn Error + Send + Sync>> + 'static,

and

> type Error = Box<dyn Error + Send + Sync>;

is common, so the service returns a boxed error and the wrapped service's error can be converted into this

Reply:
fixed 👍